File: compiler/bench/alias2.fs
\ expected: 102400000
\ Restrict-like pattern - read from one, write to another

create src 8192 allot
create dst 8192 allot

: init ( -- ) 1024 0 do i i 8 * src + ! loop ;

: copy-scaled ( -- )
  1024 0 do
    i 8 * src + @ 2 * i 8 * dst + !
  loop ;

: main
  init
  0 100000 0 do
    copy-scaled
    1024 +
  loop . cr ;

File: compiler/bench/alias2.c
// expected: 102400000
// Restrict-like pattern - read from one, write to another
#include <stdio.h>
#define SIZE 1024

long src[SIZE], dst[SIZE];

void copy_scaled(long * restrict d, const long * restrict s) {
    for (int i = 0; i < SIZE; i++) d[i] = s[i] * 2;
}

int main() {
    for (int i = 0; i < SIZE; i++) src[i] = i;
    long sum = 0;
    for (int j = 0; j < 100000; j++) {
        copy_scaled(dst, src);
        sum += SIZE;
    }
    printf("%ld\n", sum);
    return 0;
}
